pkg/kubectl/secret_for_tls.go

```diff
@@ -0,0 +1,121 @@
+type SecretForTLSGeneratorV1 struct {
+	// Name is the name of this TLS secret.
+	Name string
+	// Key is the path to the user's private key.
+	Key string
+	// Cert is the path to the user's public key certificate.
+	Cert string
+}
+
+// Ensure it supports the generator pattern that uses parameter injection
+var _ Generator = &SecretForTLSGeneratorV1{}
+
+// Ensure it supports the generator pattern that uses parameters specified during construction
+var _ StructuredGenerator = &SecretForTLSGeneratorV1{}
+
+// Generate returns a secret using the specified parameters
+func (s SecretForTLSGeneratorV1) Generate(genericParams map[string]interface{}) (runtime.Object, error) {
+	err := ValidateParams(s.ParamNames(), genericParams)
+	if err != nil {
+		return nil, err
+	}
+	params := map[string]string{}
+	for key, value := range genericParams {
+		strVal, isString := value.(string)
+		if !isString {
+			return nil, fmt.Errorf("expected string, saw %v for '%s'", value, key)
+		}
+		params[key] = strVal
+	}
+	delegate := &SecretForTLSGeneratorV1{
+		Name: params["name"],
+		Key:  params["key"],
+		Cert: params["cert"],
+	}
+	return delegate.StructuredGenerate()
+}
+
+// StructuredGenerate outputs a secret object using the configured fields
+func (s SecretForTLSGeneratorV1) StructuredGenerate() (runtime.Object, error) {
+	if err := s.validate(); err != nil {
+		return nil, err
+	}
+	tlsCrt, err := readFile(s.Cert)
+	if err != nil {
+		return nil, err
+	}
+	tlsKey, err := readFile(s.Key)
+	if err != nil {
+		return nil, err
+	}
+	secret := &api.Secret{}
+	secret.Name = s.Name
+	secret.Type = api.SecretTypeTLS
+	secret.Data = map[string][]byte{}
+	secret.Data[api.TLSCertKey] = []byte(tlsCrt)
+	secret.Data[api.TLSPrivateKeyKey] = []byte(tlsKey)
+	return secret, nil
+}
+
+// readFile just reads a file into a byte array.
+func readFile(file string) ([]byte, error) {
+	b, err := ioutil.ReadFile(file)
+	if err != nil {
+		return []byte{}, fmt.Errorf("Cannot read file %v, %v", file, err)
+	}
+	return b, nil
+}
+
+// ParamNames returns the set of supported input parameters when using the parameter injection generator pattern
+func (s SecretForTLSGeneratorV1) ParamNames() []GeneratorParam {
+	return []GeneratorParam{
+		{"name", true},
+		{"key", true},
+		{"cert", true},
+	}
+}
+
+// validate validates required fields are set to support structured generation
+func (s SecretForTLSGeneratorV1) validate() error {
+	// TODO: This is not strictly necessary. We can generate a self signed cert
+	// if no key/cert is given. The only requiredment is that we either get both
+	// or none. See test/e2e/ingress_utils for self signed cert generation.
+	if len(s.Key) == 0 {
+		return fmt.Errorf("key must be specified.")
+	}
+	if len(s.Cert) == 0 {
+		return fmt.Errorf("certificate must be specified.")
+	}
+	// TODO: Add more validation.
+	// 1. Public/private key pair match.
+	// 2. If the certificate contains intermediates, it is a valid chain.
```

Comment:
Thus validation is something @ramr has been doing with @liggitt - we should
reuse as much of that code as possible.

On Apr 25, 2016, at 3:14 PM, David Eads <notifications@github.com> wrote:

In pkg/kubectl/secret_for_tls.go
<https://github.com/kubernetes/kubernetes/pull/24719#discussion_r60948193>:

> +}
> +
> +// validate validates required fields are set to support structured generation
> +func (s SecretForTLSGeneratorV1) validate() error {
> +	// TODO: This is not strictly necessary. We can generate a self signed cert
> +	// if no key/cert is given. The only requiredment is that we either get both
> +	// or none. See test/e2e/ingress_utils for self signed cert generation.
> +	if len(s.Key) == 0 {
> +		return fmt.Errorf("key must be specified.")
> +	}
> +	if len(s.Cert) == 0 {
> +		return fmt.Errorf("certificate must be specified.")
> +	}
> +	// TODO: Add more validation.
> +	// 1. Public/private key pair match.
> +	// 2. If the certificate contains intermediates, it is a valid chain.

I could see this as a followup if you wanted to.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/24719/files/0980bbcac898a76e5bcde101615040924189a15c#r60948193>